### ISSUE TITLE
Inconsistent copyto!(sparse, sparse) and copyto!(dense, sparse)

### ISSUE THREAD

**Issue by pablosanjose** [JuliaLang/julia#34025]
```
julia> copyto!(zeros(Int, 4, 4), sparse(reshape(1:9, 3, 3)))
4×4 Array{Int64,2}:
 1  4  7  0
 2  5  8  0
 3  6  9  0
 0  0  0  0

julia> copyto!(sparse(zeros(Int, 4, 4)), sparse(reshape(1:9, 3, 3))) |> Matrix
4×4 Array{Int64,2}:
 1  5  9  0
 2  6  0  0
 3  7  0  0
 4  8  0  0
```

Which should be the correct answer?

XRef: https://github.com/JuliaLang/julia/pull/33589#issuecomment-562095892

**Comment by mcabbott** [comment]
That’s weird. The second is the behaviour for dense arrays, ` copyto!(zeros(Int, 4, 4), reshape(1:9, 3, 3)) `. I thought it always went by linear indexing, but did not know that the destinatoin could be larger, e.g. ` copyto!(zeros(Int, 12), reshape(1:9, 3, 3)') `.

**Comment by pablosanjose** [comment]
It is a bit confusing. The generic `copyto!` method between `AbtractArray`s of equal dimension uses `eachindex(src)` to iterate over the destination array, so it uses `CartesianIndices`, not linear indices. This is in https://github.com/JuliaLang/julia/blob/6f690679cc8c04b4d3b5f4e840b19524840afa97/base/multidimensional.jl#L958
However, if the two are `Array`s of different dimensions, there is a fallback that uses linear indexing in https://github.com/JuliaLang/julia/blob/6f690679cc8c04b4d3b5f4e840b19524840afa97/base/array.jl#L301

I'm actually unsure what the intended behavior is in general. The docstring is ambiguous, although it seems to hint at linear indexing
```
  copyto!(dest::AbstractArray, src) -> dest

  Copy all elements from collection src to array dest, whose length must be greater than
  or equal to the length n of src. The first n elements of dest are overwritten, the
  other elements are left untouched.
```

**Comment by mcabbott** [comment]
My example with a transpose points to `copyto!(dest::AbstractArray, src::AbstractArray)` 
https://github.com/JuliaLang/julia/blob/master/base/abstractarray.jl#L799-803
which explicitly goes by `LinearIndices(dest), LinearIndices(src)`.

**Comment by pablosanjose** [comment]
Ah, true, that method uses linear indexing despite `IndexStyle(IndexStyle(reshape(1:9, 3, 3)')) == IndexCartesian()`. So would you say that should be the general behaviour? This affects the PR #33589 I'm working on, hence my interest.

**Comment by dkarrasch** [comment]
I'd actually say the docstring is pretty clear: no requirement on dimension, just that there is "enough room" to write `src` into `dest`. As I said in #33589, one can get the "block-ish" behavior via the `copyto!` method which asks for source and dest ranges of indices, while it seems in general impossible the other way around.

**Comment by pablosanjose** [comment]
Indeed, it seems that was the intent throughout Base except for the [multidimensional.jl](https://github.com/JuliaLang/julia/blob/6f690679cc8c04b4d3b5f4e840b19524840afa97/base/multidimensional.jl#L958) method mentioned above. As I just wrote in the PR #34031, it seems the precise behavior should be to do `dest[i] = src[i]` for all `i in LinearIndices(src)`, as in the generic [abstractarray.jl](https://github.com/JuliaLang/julia/blob/master/base/abstractarray.jl#L799-803) method you linked

**Comment by timholy** [comment]
Closed by #34049
